crates/samples/components/json_validator/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+use jsonschema::JSONSchema;
+use windows::{core::*, Win32::Foundation::*};
+
+// Creates a JSON validator object with the given schema. The returned handle must be freed
+// by calling `CloseJsonValidator`.
+#[no_mangle]
+extern "system" fn CreateJsonValidator(
+    schema: *const u8,
+    schema_len: usize,
+    handle: *mut usize,
+) -> HRESULT {
+    create_validator(schema, schema_len, handle).into()
+}
+
+// Validates a JSON value against a previously-compiled schema.
+#[no_mangle]
+extern "system" fn ValidateJson(handle: usize, value: *const u8, value_len: usize) -> HRESULT {
+    validate(handle, value, value_len).into()
+}
+
+// Closes a JSON validator object.
+#[no_mangle]
+extern "system" fn CloseJsonValidator(handle: usize) {
+    if handle != 0 {
+        unsafe {
+            _ = Box::from_raw(handle as *mut JSONSchema);
+        }
+    }
+}
+
+// Implementation of the `CreateJsonValidator` function so we can use `Result` for simplicity.
+fn create_validator(schema: *const u8, schema_len: usize, handle: *mut usize) -> Result<()> {
```

Comment:
This and other functions are `unsafe` because we have to take the caller's word that the pointer is readable for the given length.

Reply:
The trouble with the pointer vs. handle debate is that the pointer-ness is an implementation detail. In the next installment I'd like to show how to implement the handles in a way that they can be validated, at which point they become merely indexes into a container.

Reply:
Opaque handle values can, by chance or purposely, point to _real allocated memory_.

Reply:
Marked them all `unsafe`.